address/address.go

```diff
@@ -47,327 +77,236 @@ const (
-	if err != nil {
-		return Address{}, err
-	}
-
-	network, err := NetworkFromString(networkString)
+// String returns an address encoded as a string.
+func (a Address) String() string {
+	str, err := encode(Mainnet, a)
 	if err != nil {
-		return Address{}, err
-	}
-
-	if version != Version {
-		return Address{}, ErrUnknownVersion
+		panic(err)
 	}
-
-	return New(network, hash), nil
+	return str
 }
 
-// NewFromBytes tries to create an address from the given bytes.
-func NewFromBytes(raw []byte) (Address, error) {
-	if len(raw) != 22 {
-		return Address{}, ErrInvalidBytes
-	}
-
-	network := raw[0]
-	if network > Testnet {
-		return Address{}, ErrUnknownNetwork
-	}
+// Equal returns true if address `b` is equal to address.
+func (a Address) Equal(b Address) bool {
+	return bytes.Equal(a.Bytes(), b.Bytes())
+}
 
-	version := raw[1]
-	if version != Version {
-		return Address{}, ErrUnknownVersion
-	}
+// Unmarshal unmarshals the cbor bytes into the address.
+func (a Address) Unmarshal(b []byte) error {
+	return cbor.DecodeInto(b, &a)
+}
 
-	return New(network, raw[2:]), nil
+// Marshal marshals the address to cbor.
+func (a Address) Marshal() ([]byte, error) {
+	return cbor.DumpObject(a)
 }
 
-// ParseError checks if the given address parses as a valid filecoin address.
-func ParseError(addr string) error {
-	hrp, version, data, err := decode(addr)
+// UnmarshalJSON implements the json unmarshal interface.
+func (a *Address) UnmarshalJSON(b []byte) error {
+	in := strings.TrimSuffix(strings.TrimPrefix(string(b), `"`), `"`)
+	addr, err := decode(in)
 	if err != nil {
-		return errors.Wrap(err, "unable to decode")
-	}
-
-	if _, err := NetworkFromString(hrp); err != nil {
-		return errors.Wrap(err, "invalid network")
-	}
-
-	if version != Version {
-		return fmt.Errorf("invalid version: version=%d", version)
-	}
-
-	if len(data) != HashLength {
-		return fmt.Errorf("invalid data length: len=%d", len(data))
+		return err
 	}
-
+	*a = addr
 	return nil
 }
 
-// Empty returns true if the address is empty.
-func (a Address) Empty() bool {
-	return a == (Address{})
+// MarshalJSON implements the json marshal interface.
+func (a Address) MarshalJSON() ([]byte, error) {
+	return json.Marshal(a.String())
 }
 
-// Network returns the network of the address.
-func (a Address) Network() Network {
-	return a[0]
+// NewIDAddress returns an address using the ID protocol.
+func NewIDAddress(id uint64) Address {
+	return newAddress(ID, leb128.FromUInt64(id))
 }
 
-// Version returns the version of the address.
-func (a Address) Version() byte {
-	return a[1]
+// NewSecp256k1Address returns an address using the SECP256K1 protocol.
+func NewSecp256k1Address(pubkey []byte) Address {
+	return newAddress(SECP256K1, addressHash(pubkey))
 }
 
-// Hash returns the hash part of the address.
-func (a Address) Hash() []byte {
-	return a[2:]
+// NewActorAddress returns an address using the Actor protocol.
+func NewActorAddress(data []byte) Address {
```

Comment:
Are there any callers that don't start with a string here? Perhaps just define it that way.